src/nvim/runtime.c

```diff
@@ -227,10 +314,152 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
+        push_path(search_path, xstrdup((char *)files[i]), false);
+        size_t after_size = STRLEN(files[i])+6;
+        char *after = xmallocz(after_size);
+        xstrlcpy(after, (char *)files[i], after_size);
+        xstrlcat(after, "after/", after_size);
+        if (os_isdir((char_u *)after)) {
+          push_path(search_path, after, true);
+        } else {
+          xfree(after);
+        }
+      }
+      FreeWild(num_files, files);
+    }
+  }
+}
+
+static bool path_is_after(char_u *buf, size_t buflen)
+{
+  // NOTE: we only consider dirs exactly matching "after" to be an AFTER dir.
+  // vim8 considers all dirs like "foo/bar_after", "Xafter" etc, as an
+  // "after" dir in SOME codepaths not not in ALL codepaths.
+  return buflen >= 5
+         && (!(buflen >= 6) || vim_ispathsep(buf[buflen-6]))
+         && STRCMP(buf + buflen - 5, "after") == 0;
+}
+
+RuntimeSearchPath build_runtime_search_path(void)
+{
+  kvec_t(String) pack_entries = KV_INITIAL_VALUE;
+  Map(String, handle_T) pack_used = MAP_INIT;
+  // TODO(bfredl): add a set of existing rtp entries to not duplicate those
+  RuntimeSearchPath search_path = KV_INITIAL_VALUE;
+  CharVec after_path = KV_INITIAL_VALUE;
+
+  static char_u buf[MAXPATHL];
+  for (char *entry = (char *)p_pp; *entry != NUL; ) {
+    char *cur_entry = entry;
+    copy_option_part((char_u **)&entry, buf, MAXPATHL, ",");
+
+    String the_entry = { .data = cur_entry, .size = STRLEN(buf) };
+
+    kv_push(pack_entries, the_entry);
+    map_put(String, handle_T)(&pack_used, the_entry, 0);
+  }
+
+
+  char *rtp_entry;
+  for (rtp_entry = (char *)p_rtp; *rtp_entry != NUL; ) {
+    char *cur_entry = rtp_entry;
+    copy_option_part((char_u **)&rtp_entry, buf, MAXPATHL, ",");
+    size_t buflen = STRLEN(buf);
+
+    if (path_is_after(buf, buflen)) {
+      rtp_entry = cur_entry;
+      break;
+    }
+
+    push_path(&search_path, xstrdup((char *)buf), false);
+
+    handle_T *h = map_ref(String, handle_T)(&pack_used, cstr_as_string((char *)buf), false);
+    if (h) {
+      (*h)++;
+      expand_pack_entry(&search_path, &after_path, buf);
+    }
+  }
+
+  for (size_t i = 0; i < kv_size(pack_entries); i++) {
+    handle_T h = map_get(String, handle_T)(&pack_used, cstr_as_string((char *)buf));
+    if (h == 0) {
+      expand_pack_entry(&search_path, &after_path, (char_u *)kv_A(pack_entries, i).data);
+    }
+  }
+
+  // "after" packages
+  for (size_t i = 0; i < kv_size(after_path); i++) {
+    push_path(&search_path, kv_A(after_path, i), true);
+  }
+
+  // "after" dirs in rtp
+  for (; *rtp_entry != NUL;) {
+    copy_option_part((char_u **)&rtp_entry, buf, MAXPATHL, ",");
+    push_path(&search_path, xstrdup((char *)buf), path_is_after(buf, STRLEN(buf)));
+  }
+
+  // strings are not owned
+  kv_destroy(pack_entries);
+  map_destroy(String, handle_T)(&pack_used);
+
+  return search_path;
+}
+
+void invalidate_search_path(void)
+{
+  search_path_valid = false;
+}
+
+void validate_search_path(void)
```

Comment:
nit: would be nice to have a common prefix.

```
runtime_searchpath_validate()
runtime_searchpath_invalidate()
runtime_searchpath_build()
runtime_searchpath_push()
runtime_path_is_after()
runtime_expand_pack_entry()

Reply:
I will do it for the next installment of the series (very soon)